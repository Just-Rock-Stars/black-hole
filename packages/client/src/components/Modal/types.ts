import { ReactNode } from 'react'

export interface ModalProps {
  isOpen: boolean
  title?: string
  children: ReactNode
  onClose: VoidFunction
}
