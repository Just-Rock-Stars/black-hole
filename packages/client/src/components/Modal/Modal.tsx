import { FC } from 'react';
import { createPortal } from 'react-dom';
import { FiX } from 'react-icons/fi';

import { Background } from '@components/Background';

import { TModalProps } from './types';

export const Modal: FC<TModalProps> = ({ isOpen, title, children, onClose }) => {
  if (!isOpen) return null;

  return createPortal(
    <div className="fixed flex flex-col inset-0 bg-black p-4">
      <Background />
      <div className="relative z-1 flex justify-end">
        {title && <header className="mr-auto text-black">{title}</header>}
        <button onClick={onClose}>
          <FiX className="w-8 h-8 text-white" />
        </button>
      </div>
      <div className="relative z-1 flex items-center justify-center grow">{children}</div>
    </div>,
    document.body
  );
};
