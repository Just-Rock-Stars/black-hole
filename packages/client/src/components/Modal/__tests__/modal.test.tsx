import { fireEvent, render, screen } from '@testing-library/react';

import { Modal } from '../Modal';

const onClose = jest.fn();

describe('Modal snapshots', () => {
  test('should render modal', () => {
    const tree = render(
      <Modal isOpen={true} onClose={onClose}>
        <h1>Modal children</h1>
      </Modal>
    );

    expect(tree).toMatchSnapshot();
  });

  test('should not render modal', () => {
    const tree = render(
      <Modal isOpen={false} onClose={onClose}>
        <h1>Modal children</h1>
      </Modal>
    );

    expect(tree).toMatchSnapshot();
  });
});

describe('Modal behavior', () => {
  test('should render title', () => {
    render(
      <Modal isOpen={true} onClose={onClose} title="Modal title">
        <h1>Modal children</h1>
      </Modal>
    );

    const title = screen.getByText('Modal title');

    expect(title).toBeInTheDocument();
  });

  test('should click on close button', () => {
    render(
      <Modal isOpen={true} onClose={onClose}>
        <h1>Modal children</h1>
      </Modal>
    );

    const closeButton = screen.getByRole('button');

    fireEvent.click(closeButton);

    expect(onClose).toBeCalledTimes(1);
  });
});
