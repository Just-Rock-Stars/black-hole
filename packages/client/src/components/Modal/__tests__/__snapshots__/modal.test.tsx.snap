// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal snapshots should not render modal 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Modal snapshots should render modal 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div
      class="fixed flex flex-col inset-0 bg-white p-4"
    >
      <div
        class="flex justify-end"
      >
        <button>
          <svg
            class="w-8 h-8"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </button>
      </div>
      <div
        class="flex items-center justify-center grow"
      >
        <h1>
          Modal children
        </h1>
      </div>
    </div>
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
